src/erf.jl

```diff
@@ -4,40 +4,73 @@ using Base.Math: @horner
-        ($f)(x::Real) = ($f)(float(x))
-        ($f)(a::Float16) = Float16($f(Float32(a)))
-        ($f)(a::Complex{Float16}) = Complex{Float16}($f(Complex{Float32}(a)))
-        function ($f)(x::BigFloat)
+        $f(x::Number) = $internalf(float(x))
+
+        $internalf(x::Float64) = ccall(($libopenlibmf, libopenlibm), Float64, (Float64,), x)
+        $internalf(x::Float32) = ccall(($libopenlibmf0, libopenlibm), Float32, (Float32,), x)
+        $internalf(x::Float16) = Float16($internalf(Float32(x)))
+
+        $internalf(z::Complex{Float64}) = Complex{Float64}(ccall(($openspecfunf, libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), z, zero(Float64)))
+        $internalf(z::Complex{Float32}) = Complex{Float32}(ccall(($openspecfunf, libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), Complex{Float64}(z), Float64(eps(Float32))))
+        $internalf(z::Complex{Float16}) = Complex{Float16}($internalf(Complex{Float32}(z)))
+
+        function $internalf(x::BigFloat)
             z = BigFloat()
-            ccall(($(string(:mpfr_,f)), :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, Int32), z, x, ROUNDING_MODE[])
+            ccall(($mpfrf, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, Int32), z, x, ROUNDING_MODE[])
             return z
         end
-        ($f)(x::AbstractFloat) = error("not implemented for ", typeof(x))
     end
 end
 
-for f in (:erf, :erfc, :erfcx, :erfi, :Dawson)
-    fname = (f === :Dawson) ? :dawson : f
+for f in (:erfcx, :erfi, :dawson)
+    internalf = Symbol(:_, f)
+    openspecfunfsym = Symbol(:Faddeeva_, f === :dawson ? :Dawson : f)
+    openspecfunfF64 = QuoteNode(Symbol(openspecfunfsym, :_re))
+    openspecfunfCF64 = QuoteNode(openspecfunfsym)
     @eval begin
-        ($fname)(z::Complex{Float64}) = Complex{Float64}(ccall(($(string("Faddeeva_",f)),libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), z, zero(Float64)))
-        ($fname)(z::Complex{Float32}) = Complex{Float32}(ccall(($(string("Faddeeva_",f)),libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), Complex{Float64}(z), Float64(eps(Float32))))
+        $f(x::Number) = $internalf(float(x))
+
+        $internalf(x::Float64) = ccall(($openspecfunfF64, libopenspecfun), Float64, (Float64,), x)
+        $internalf(x::Float32) = Float32($internalf(Float64(x)))
+        $internalf(x::Float16) = Float16($internalf(Float64(x)))
 
-        ($fname)(z::Complex) = ($fname)(float(z))
-        ($fname)(z::Complex{<:AbstractFloat}) = throw(MethodError($fname,(z,)))
+        $internalf(z::Complex{Float64}) = Complex{Float64}(ccall(($openspecfunfCF64, libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), z, zero(Float64)))
+        $internalf(z::Complex{Float32}) = Complex{Float32}(ccall(($openspecfunfCF64, libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), Complex{Float64}(z), Float64(eps(Float32))))
+        $internalf(z::Complex{Float16}) = Complex{Float16}($internalf(Complex{Float32}(z)))
```

Comment:
This adds support for `Complex{Float16}` in the same way as already present for `erf` etc. above. I can move it to a separate PR if desired.